pkiinit/option/cache.go

```diff
@@ -0,0 +1,80 @@
+//
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+//
+// SPDX-License-Identifier: Apache-2.0'
+//
+package option
+
+import (
+	"os"
+	"path/filepath"
+)
+
+// Cache option....
+func Cache() func(*PkiInitOption) (exitCode, error) {
+	return func(pkiInitOpton *PkiInitOption) (exitCode, error) {
+
+		if isCacheNoOp(pkiInitOpton) {
+			return normal, nil
+		}
+
+		return cachePkis(false)
+	}
+}
+
+func isCacheNoOp(pkiInitOption *PkiInitOption) bool {
+	// nop: if the flag is missing or not on
+	return pkiInitOption == nil || !pkiInitOption.CacheOpt
+}
+
+func cachePkis(cacheca bool) (statusCode exitCode, err error) {
+	// first perform generate func
+	statusCode, err = generatePkis()
+	if err != nil {
+		return statusCode, err
+	}
+
+	// not caching the CA private key
+	if err = doCache(!cacheca); err != nil {
+		return exitWithError, err
+	}
+
+	// to deploy
+	// copy stuff into dest dir from pkiCache
+	err = deploy(getPkiCacheDirEnv(), pkiInitDeployDir)
+	if err != nil {
+		return exitWithError, err
+	}
+
+	return normal, nil
+}
+
+func doCache(eraseCAPrivateKey bool) error {
+	// source
+	generatedDirPath := filepath.Join(os.Getenv(envXdgRuntimeDir), pkiInitGeneratedDir)
+
+	if eraseCAPrivateKey {
+		// shreds CA private key before copy to cache dir
+		os.Remove(filepath.Join(generatedDirPath, caServiceName, tlsSecretFileName))
```

Comment:
This is not sufficient to securely erase the CA private key.  You need to open the file and do random I/O to overwrite every byte in the file with zeroes, then remove the file.

Reply:
You will also want to security erase the CA private key (with -generate) even if you aren't caching the PKI (unless you are explicitly caching the CA)

Reply:
ok